File: lib/presentation/widgets/app_avatar.dart
import 'package:flutter/material.dart';

/// A generic avatar that shows a network image or falls back to the
/// user's initials, with an optional online status dot.
class AppAvatar extends StatelessWidget {
  const AppAvatar({
    super.key,
    required this.name,
    this.imageUrl,
    this.radius = 28,
    this.online,
  });

  final String name;
  final String? imageUrl;
  final double radius;

  /// null = no status dot, true = green, false = grey.
  final bool? online;

  String get _initials {
    final parts = name.trim().split(RegExp(r'\s+'));
    if (parts.isEmpty || parts.first.isEmpty) return '?';
    if (parts.length == 1) return parts.first.characters.first.toUpperCase();
    return (parts.first.characters.first + parts.last.characters.first)
        .toUpperCase();
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    return Stack(
      children: [
        CircleAvatar(
          radius: radius,
          backgroundColor: theme.colorScheme.primaryContainer,
          backgroundImage: imageUrl != null ? NetworkImage(imageUrl!) : null,
          child: imageUrl == null
              ? Text(
                  _initials,
                  style: TextStyle(
                    fontSize: radius * 0.7,
                    fontWeight: FontWeight.bold,
                    color: theme.colorScheme.onPrimaryContainer,
                  ),
                )
              : null,
        ),
        if (online != null)
          Positioned(
            right: 0,
            bottom: 0,
            child: Container(
              width: radius * 0.5,
              height: radius * 0.5,
              decoration: BoxDecoration(
                color: online! ? Colors.green : Colors.grey,
                shape: BoxShape.circle,
                border: Border.all(color: theme.colorScheme.surface, width: 2),
              ),
            ),
          ),
      ],
    );
  }
}

File: android_native/app/src/main/java/com/example/widgetgallery/presentation/components/AppAvatar.kt
package com.example.widgetgallery.presentation.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

internal fun initialsOf(name: String): String {
    val p = name.trim().split(Regex("\\s+"))
    if (p.isEmpty() || p[0].isEmpty()) return "?"
    return if (p.size == 1) p[0].first().uppercase()
    else (p.first().first().toString() + p.last().first()).uppercase()
}

/** Avatar with initials fallback and an optional online status dot. */
@Composable
fun AppAvatar(
    name: String,
    modifier: Modifier = Modifier,
    radius: Dp = 28.dp,
    online: Boolean? = null,
) {
    val scheme = MaterialTheme.colorScheme
    Box(modifier.size(radius * 2)) {
        Box(
            Modifier.size(radius * 2).clip(CircleShape)
                .background(scheme.primaryContainer),
            contentAlignment = Alignment.Center,
        ) {
            Text(initialsOf(name), color = scheme.onPrimaryContainer,
                fontWeight = FontWeight.Bold,
                fontSize = (radius.value * 0.7f).sp)
        }
        if (online != null) {
            Box(
                Modifier
                    .align(Alignment.BottomEnd)
                    .size(radius * 0.55f)
                    .clip(CircleShape)
                    .background(scheme.surface),
                contentAlignment = Alignment.Center,
            ) {
                Box(
                    Modifier.size(radius * 0.4f).clip(CircleShape)
                        .background(
                            if (online) Color(0xFF2E7D32)
                            else Color(0xFF9E9E9E)
                        )
                )
            }
        }
    }
}
